[WIP] Terraform Docs for TF policy (#2690)
hashicorp/web-unified-docs: content/terraform-policy/v0.1.x (beta)/docs/policy/write/policy-blocks.mdx

@include "partials/beta.mdx"

## Policy block syntax

Terraform policy block syntax is based on HCL, and similar to Terraform's syntax.

```hcl
<block_type> "<label>" ["<name>"] {
  <attribute_name> = <attribute_value>
  ...

  <block_name> {
    <block_attr_name> = <block_attr_value>
    ...
  }
}
```

Each policy file can include:

- An optional ['policy {}' block](#the-policy--block) to configure Terraform policy itself.
- One or more [`provider_policy {}`](#the-provider-policy--block), [`resource_policy {}`](#the-resource-policy--block), and [`module_policy {}`](#the-module-policy--block) blocks to define your policies.
- Any number of [`input {}`](/terraform/policy/write/inputs-and-locals#the-input--block) blocks to define input variables.
- Any number of [`locals {}`](/terraform/policy/write/inputs-and-locals#the-locals--block) blocks to define local values.

Terraform policy will evaluate all polices in a policy file as a unit, as early in the Terraform run as possible, when all values that the given policy block refers to are available.

> ## The `policy {}` block

Reviewer: ```suggestion
## `policy`
```
For all of these, the heading should match the keyword as the practitioner would type it. We should be following the [language reference type guidance](https://hashicorp.atlassian.net/wiki/spaces/TW/pages/2675737045/Template+for+structured+configuration+reference)

> Use the policy block to configure how Terraform policy evaluates your policies, including the version of Terraform that Terraform policy will use to evaluate your policies.

Reviewer: ```suggestion
The `policy` block configures how Terraform policy evaluates your policies, including which version of Terraform to use to perform the evaluation.
```

> The `policy` block supports the optional `terraform_config` attribute, which defines settings which are specific to Terraform. The only currently supported attribute is `required_version`, which validates that the current Terraform CLI version fulfills the specified version constraint. The `required_version` attribute uses the same version constraint syntax as Terraform itself. Refer to [Version constraints - Version constraint syntax](/terraform/language/expressions/version-constraints#version-constraint-syntax) for more information.

Reviewer: This says that you can add the `terraform_config` attribute, but the next sentence says that the only attribute the block supports is `required_version`. Which is it?

The `policy` block also supports the optional `plugins` attribute, which allows you to refer to binary plugins that provide custom functions you can use in your policies. Refer to [Terraform policy - Policy plugins](/terraform/policy/write/plugins) for more information.

```hcl
policy {
  terraform_config {
    required_version = "> 1.16.0"
  }
  plugins {
    example = {
      source = "../plugins/bin/example"
    }
  }
}
```

The above example requires that the Terraform version used to evaluate these policies be greater than "1.16.0".

> ## Policy blocks

Reviewer: ??? - Didn't we just describe the `policy` block?

Terraform policy supports three types of block that allow you to write policies for your Terraform providers, resources, and modules.

> Each policy block requires two labels. The first label will match the target provider, resource, or module, and the second gives the policy a unique name within the same policy file. Optionally, you can use `*` for all or part the first label to match multiple providers, resources, or modules in your Terraform configuration.

Reviewer: You are describing the `<construct> "<type>" "<label>"` pattern, correct? 

This content should follow the format we use for all construct reference docs, such as:

https://developer.hashicorp.com/terraform/language/block/tfquery/list#list

https://developer.hashicorp.com/terraform/language/block/action#action 

And here's an example implementation in Consul:
https://developer.hashicorp.com/consul/docs/reference/config-entry/api-gateway

By default, Terraform policy will apply a given policy to all of the providers, resources, or modules that match the given label. You can filter the matching providers, resources, or modules with the [`filter` meta-argument](/terraform/policy/write/attributes#the-filter-meta-argument).

By default, Terraform policy will treat your policies as advisory. You can set the enforcement level with the optional [`enforcement_level` meta-argument](/terraform/policy/write/attributes#the-enforcement_level-meta-argument). There are three enforcement levels available: "advisory", "mandatory-overridable" and "mandatory".

By default, Terraform policy will enforce your policy as early in a Terraform operation as possible. You can specify the phases during which Terraform policy will evaluate your policy with the optional [`operations` meta-argument](/terraform/policy/write/attributes#the-operations-meta-argument).

> Each policy block must contain one or more `enforce {}` blocks that define a condition that must evaluate to `true` for the policy to pass. Refer to [The `enforce {}` block](#the-enforce--block) for more informaiton.

Reviewer: Here is what I've been able to figure out:

Define your policy in a `.policy.hcl` file

This file supports the following constructs and arguments (which we are calling attributes in this syntax, correct?):

`policy`
 - `terraform_config`
   - `required_version`: <version constraint>
 - `plugins`
    - <plugin-name> 
      - `source`

`<unclear-construct-keyword> "<provider>" "<label>"`
- `filter` 
- `enforcement_level`
- `operations`
- `enforce`

I _think_ the <unclear-construct-keyword> refers to `provider_policy`, `resource_policy`, and `module_policy`.
